### PR TITLE
feat: PartnerRegistrar

### PR DESCRIPTION
**Pull request by OnyekaIjeh**

### CODE REVIEW

File: src/PartnerRegistrar.ts [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+   */
+  getProvider (): providers.JsonRpcProvider {
+    return this.jsonRpcProvider
+  }
+
+  /**
+   * Returns the signer
+   * @param signer will just return the signer if provided, otherwise will return the (signer) of the class
+   */
+  getSigner (signer?: Signer): Signer {
```
Comment (sergio-jara) [inline src/PartnerRegistrar.ts]: is it possible to make it private? I dont see the need to expose it.

File: src/PartnerRegistrar.ts [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+  private readonly partnerRegistrar: Contract
+  private readonly rifToken: Contract
+  private readonly jsonRpcProvider: providers.JsonRpcProvider
+
+  constructor (
+    partnerRegistrarAddress: string,
+    private readonly partnerAddress: string,
+    rskOwnerAddress: string,
+    rifTokenAddress: string,
+    provider: string,
```
Comment (sergio-jara) [inline src/PartnerRegistrar.ts]: can we justify including this paremeter given that the provider is also included in the signer object?

File: src/PartnerRegistrar.ts [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+   * @param owner the owner of the name
+   * @param duration the duration to register the name
+   * @param signer the signer for the transaction
+   * @param addr the address to set for the name resolution
+   */
+  async commit (label: string, owner: string, duration: BigNumber, signer?: Signer, addr?: string): Promise<{ secret: string, hash: string }> {
+    const _signer = this.getSigner(signer)
+    const secret = generateSecret()
+    const hash = await this.partnerRegistrar.makeCommitment(hashLabel(label), owner, secret, duration, addr ?? owner)
+    const makeCommitmentTransaction = await this.partnerRegistrar.connect(_signer).commit(hash, this.partnerAddress)
```
Comment (sergio-jara) [inline src/PartnerRegistrar.ts]: doesnt the provider set in the constructor get overridden here?

File: src/PartnerRegistrar.ts [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+  }
+
+  /**
+   * Make a commitment
+   * @param label the name to register
+   * @param owner the owner of the name
+   * @param duration the duration to register the name
+   * @param signer the signer for the transaction
+   * @param addr the address to set for the name resolution
+   */
```
Comment (sergio-jara) [inline src/PartnerRegistrar.ts:79]: docs description for the returned values? what is secret? what is hash?

File: src/PartnerRegistrar.ts [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+    const makeCommitmentTransaction = await this.partnerRegistrar.connect(_signer).commit(hash, this.partnerAddress)
+    await makeCommitmentTransaction.wait()
+
+    return {
+      secret,
+      hash
+    }
+  }
+
+  async canReveal (hash: string): Promise<boolean> {
```
Comment (sergio-jara) [inline src/PartnerRegistrar.ts:96]: docs?

Comment (sergio-jara) [reply src/PartnerRegistrar.ts:96]: steal it from the contracts lol

File: src/PartnerRegistrar.ts [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+    const _name = Buffer.from(utils.toUtf8Bytes(label)).toString('hex')
+
+    const data = `${_signature}${_owner}${_secret}${_duration}${_addr}${_partner}${_name}`
+
+    const _signer = this.getSigner(signer)
+
+    const transaction = await this.rifToken.connect(_signer).transferAndCall(this.partnerRegistrar.address, amount, data)
+
+    return transaction.wait()
+  }
```
Comment (sergio-jara) [inline src/PartnerRegistrar.ts:133]: excelent

File: src/PartnerRegistrar.ts [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+   * @param label the name to register
+   * @param owner the owner of the name
+   * @param duration the duration to register the name
+   * @param amount the amount for the name registration
+   * @param partnerConfigurationAddress the address of the partner configuration contract
+   * @param addr the address to set for the name resolution
+   * @param signer the signer for the transaction
+   */
+  async commitAndRegister (label: string, owner: string, duration: BigNumber, amount: BigNumber, partnerConfigurationAddress: string, addr?: string, signer?: Signer): Promise<boolean> {
+    const REVEAL_TIMEOUT_BUFFER = 30 * 1000 // 30 seconds (time to mine a block in RSK)
```
Comment (sergio-jara) [inline src/PartnerRegistrar.ts:145]: this value being hardcoded... 👀
I know... but still
maybe we can define it as a param at the end of this signature and make it optional with a default value of 30. how bout that?

File: src/PartnerRegistrar.ts [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+        }, minCommitmentAge.toNumber() * 1000 + REVEAL_TIMEOUT_BUFFER)
+      })
+
+      if (!canReveal) {
+        throw new Error('Cannot register because the commitment cannot be revealed')
+      }
+    }
+
+    return this.register(label, owner, secret, duration, amount, addr, _signer)
+  }
```
Comment (sergio-jara) [inline src/PartnerRegistrar.ts:174]: this is great work.

File: test/rskPartnerRegistrar.test.ts [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+      } = await partnerRegistrar.commit(name, rnsOwnerAddress, duration, rnsOwner)
+
+      expect(await partnerRegistrar.canReveal(hash)).toBe(false)
+
+      await timeTravel(provider, defaultMinCommitmentAge)
+
+      expect(await partnerRegistrar.canReveal(hash)).toBe(true)
+    }, 3000000)
+  })
+})
```
Comment (sergio-jara) [inline test/rskPartnerRegistrar.test.ts:335]: damn. awesome

**Review by sergio-jara — COMMENTED** [review]
Great work. 
Left a few comments to discuss them.

File: src/PartnerRegistrar.ts (hunk 2 above)
Comment (OnyekaIjeh) [reply src/PartnerRegistrar.ts]: The signer is not required, but the provider is required to read.

File: src/PartnerRegistrar.ts (hunk 3 above)
Comment (OnyekaIjeh) [reply src/PartnerRegistrar.ts]: The provider is required to instantiate the instance of the contract.
I may be wrong about this whole provider and signer thing. We can talk about it.

File: src/PartnerRegistrar.ts (hunk 4 above)
Comment (OnyekaIjeh) [reply src/PartnerRegistrar.ts:115]: will add it

File: src/PartnerRegistrar.ts (hunk 1 above)
Comment (OnyekaIjeh) [reply src/PartnerRegistrar.ts:75]: done

File: src/PartnerRegistrar.ts (hunk 5 above)
Comment (OnyekaIjeh) [reply src/PartnerRegistrar.ts:129]: done

File: src/PartnerRegistrar.ts (hunk 7 above)
Comment (OnyekaIjeh) [reply src/PartnerRegistrar.ts:145]: hmmm

Comment (OnyekaIjeh) [reply src/PartnerRegistrar.ts:145]: not sure about that one, it is pretty much a constant

**Review by sergio-jara — APPROVED** [review]

**Pull request merged**